### PR TITLE
AVRO-3547: Add support for custom attributes in Avro Schema

### PR DESCRIPTION
**Pull request by nileyadav**
Custom attributes are allowed at field level as per Avro specification : https://avro.apache.org/docs/current/spec.html#schemas

But Avro c++ implementation does not support custom attribute at field level.
Update implementation to support custom attribute at field level.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3547
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests:
  - Added test cases for custom fields under 
    - avro::schema::testBasic
    - avro::schema::testCompile
    - avro::Schema::testRoundTrip 
  - Added units to validate Node record and schema build in unittest.cc

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by emkornfield** [comment]
Sorry if I missed it but are there unit tests associated with this change?

**Comment by nileyadav** [comment]
> 

I missed 2 files, have added those in 2nd commit.

File: lang/c++/impl/CustomFields.cc
Comment (martin-g) [inline lang/c++/impl/CustomFields.cc:24]:
```suggestion
#include "Exception.hh"

```

File: lang/c++/impl/CustomFields.cc
Comment (martin-g) [inline lang/c++/impl/CustomFields.cc:25]:
```suggestion
namespace avro {

```

File: lang/c++/impl/CustomFields.cc
Comment (martin-g) [inline lang/c++/impl/CustomFields.cc:34]:
```suggestion
}

```

File: lang/c++/impl/CustomFields.cc
Comment (martin-g) [inline lang/c++/impl/CustomFields.cc:47]:
```suggestion
}

```

File: lang/c++/impl/NodeImpl.cc
Comment (martin-g) [inline lang/c++/impl/NodeImpl.cc:91]:
```suggestion
      os << "," << endl  << indent(depth);
```

File: lang/c++/impl/NodeImpl.cc
Comment (martin-g) [inline lang/c++/impl/NodeImpl.cc:292]: Should this use `endl` instead ?

**Comment by emkornfield** [comment]
@nileyadav it looks like the CI is failing with compilation errors

**Comment by nileyadav** [comment]
> @nileyadav it looks like the CI is failing with compilation errors

I've fixed the build issues

**Review by martin-g — APPROVED** [review]

**Comment by nileyadav** [comment]
> @martin-g Thank you for review! I do not have write access to apache::master, could you please help me by merging this PR to apache::master?

Gentle ping!

**Comment by martin-g** [comment]
I hope someone else will also review the changes.
I am refreshing my C++ knowledge here, but I haven't coded in C++ for 15 years ...

File: lang/c++/impl/Compiler.cc
Comment (nileyadav) [reply lang/c++/impl/Compiler.cc:153]: Exactly, it'll just make it const&&.

File: lang/c++/api/CustomFields.hh
Comment (emkornfield) [inline lang/c++/api/CustomFields.hh:28]: please add some docs, on what CustomFields are.

File: lang/c++/impl/Compiler.cc
Comment (emkornfield) [inline lang/c++/impl/Compiler.cc:277]: !kKnownFields.contains(...) might be easier to read.

Comment (emkornfield) [reply lang/c++/impl/Compiler.cc:277]: actually this is C++20 so probably not portable.

**Review by emkornfield — APPROVED** [review]

File: lang/c++/api/CustomFields.hh
Comment (nileyadav) [reply lang/c++/api/CustomFields.hh:31]: Added documentation for class

**Comment by nileyadav** [comment]
@KalleOlaviNiemitalo please let me know if you have any other comments

**Comment by nileyadav** [comment]
@martin-g the PR has been approved by [emkornfield](https://github.com/emkornfield) as well.

**Comment by emkornfield** [comment]
Just a note, I'm not a committer.  Is there an committer more familiar with the C++ codebase that we can ask for a review?

**Comment by nileyadav** [comment]
> I hope someone else will also review the changes.
> I am refreshing my C++ knowledge here, but I haven't coded in C++ for 15 years ...

Hello @martin-g , would you please suggest an committer whom I should ask for a review, before this PR could be merged?

**Comment by martin-g** [comment]
Thank you for the contribution, @nileyadav !

Thank you for the reviews, @KalleOlaviNiemitalo and @emkornfield !

**Comment by nileyadav** [comment]
> Thank you for the contribution, @nileyadav !
> 
> Thank you for the reviews, @KalleOlaviNiemitalo and @emkornfield !

Thank you @martin-g , @emkornfield and @KalleOlaviNiemitalo for helping me with my contribution to Avro library.

**Comment by mruffing** [comment]
@nileyadav This include https://github.com/apache/avro/blob/master/lang/c%2B%2B/api/CustomFields.hh#L24 (introduced in this PR) breaks the API headers that are installed using the `build.sh install` script.  Specifically `"../impl/json/JsonDom.hh"` is not in the include folder (generated by `build.sh install`).

**Comment by martin-g** [comment]
Thanks for the report, @mruffing !
Please create a new JIRA for the problem!
A PR with a fix would be awesome!

**Comment by chenrui333** [comment]
> Specifically "../impl/json/JsonDom.hh" is not in the include folder (generated by build.sh install).

running the same issue with avro-cpp build.

**Comment by thiru-mg** [comment]
I'm sorry I didn't look at the PR before. I think the approach taken here exposes the `json` namespace as part of public API. Any future changes hereafter will have to be backwards compatible. When I originally designed the `json` support I took several decisions freely because it was private. Now in one stroke the whole thing opens up for the public.
Even though support for custom-fields is indeed a welcome enhancement, the side effect of exposing json is much bigger decision. I recommend we revert this change and reintroduce it without exposing json or after a complete review of the JSON.
Again, I'm sorry I failed to notice it early enough.

**Comment by thiru-mg** [comment]
Here is one possible solution, let's make the custom attribute `map<string, string>` The key of the map is the name of the custom attribute and the value is the JSON encoded string. It is up to the users to parse and interpret the json, using their favorite JSON parser. We don't have to force our JSON on them or get exposed.

**Comment by thiru-mg** [comment]
By the way, I also suggest renaming `CustomFields` to `customAttributes`, the term `Field` makes it appear to be related to field used in Avro record schemata.

**Comment by martin-g** [comment]
I will work on this approach!

**Comment by chenrui333** [comment]
tag me once I have some patch to try on with my homebrew upgrade PR, thanks!

**Comment by martin-g** [comment]
@chenrui333 Current `master` should be OK for your build.
https://github.com/apache/avro/pull/1826 brings some more improvements too.

**Comment by chenrui333** [comment]
Thanks @martin-g gonna give it a shot.

**Comment by chenrui333** [comment]
Works for me now, retry in [this PR](https://github.com/Homebrew/homebrew-core/pull/118877)

**Pull request merged**
